synapse/neuron.py

```diff
@@ -447,91 +445,179 @@ def initConfDefs(self):
+        self.chain(link)
+        self._sess_boss = boss
+        self.is_lisn = lisn    # True if we are the listener.
+        self._crypter = None  # type: CryptSeq
+        self._my_ephem_prv = None  # type: s_ecc.PriKey
 
     def handlers(self):
         return {
-            'cert': self._onMesgCert,
-            'skey': self._onMesgSkey,
+            'helo': self._onMesgHelo,
             'xmit': self._onMesgXmit,
+            'fail': self._onMesgFail
         }
 
-    def setRxKey(self, rxkey):
-        self.rxkey = rxkey
-        self.rxtinh = s_tinfoil.TinFoilHat(rxkey)
-
     def _tx_real(self, mesg):
 
-        if self.txtinh is None:
+        if self._crypter is None:
             raise s_exc.NotReady()
 
-        data = self.txtinh.enc(s_msgpack.en(mesg))
+        data = self._crypter.encrypt(mesg)
         self.link.tx(('xmit', {'data': data}))
 
+    def _onMesgFail(self, link, mesg):
+        logger.error('Remote peer issued error: %r.', mesg)
+        self.fini()
+
     def _onMesgXmit(self, link, mesg):
 
-        if self.rxtinh is None:
-            logger.warning('xmit message before rxkey')
+        if self._crypter is None:
+            logger.warning('xmit message before session establishment complete')
             raise s_common.NotReady()
 
-        data = mesg[1].get('data')
-        newm = s_msgpack.un(self.rxtinh.dec(data))
+        ciphertext = mesg[1].get('data')
+        newm = self._crypter.decrypt(ciphertext)
 
         try:
             self.taskplex.rx(self, newm)
         except Exception as e:
             logger.exception('xmit taskplex error')
+            self.fini()
 
-    def _onMesgSkey(self, link, mesg):
+    def _send_fail(self, exc, exc_info=None):
+        self.link.tx(('fail', {'exception': (repr(exc), exc_info)}))
 
-        data = mesg[1].get('data')
+    @s_glob.inpool
+    def _initiateSession(self):
+        ''' (As the initiator) start a new session
 
-        skey = self._sess_boss.decrypt(data)
+        Send ephemeral public and my certificate
+        '''
+        if self.is_lisn:
+            raise Exception('Listen link cannot initiate a session')
+        self._my_ephem_prv = s_ecc.PriKey.generate()
+        self.link.tx(('helo', {'version': NEURON_PROTO_VERSION,
+                               'ephem_pub': self._my_ephem_prv.public().dump(),
+                               'cert': self._sess_boss.certbyts}))
 
-        self.setRxKey(skey)
+    def _handle_session_msg(self, mesg):
+        ''' Validate and set up the crypto from a helo message '''
+        if self._crypter:
+            raise ProtoError('Received two client helos')
 
-    def sendcert(self):
-        self.link.tx(('cert', {'cert': self._sess_boss.certbyts}))
+        if self.is_lisn:
+            self._my_ephem_prv = s_ecc.PriKey.generate()
 
-    @s_glob.inpool
-    def _onMesgCert(self, link, mesg):
+        version = mesg[1].get('version')
+        if version != NEURON_PROTO_VERSION:
```

Comment:
Do we only want to validate the first number (major version) match here? in the event that we may want to do a minor version rev?